includes/wp-admin/class-admin.php

```diff
@@ -593,23 +598,207 @@ public static function user_bulk_options( $actions ) {
 	public static function handle_bulk_request( $send_back, $action, $users ) {
 		if (
 			'remove_activitypub_cap' !== $action &&
-			'add_activitypub_cap' !== $action
+			'add_activitypub_cap' !== $action &&
+			'remove_activitypub_cap_confirmed' !== $action
 		) {
 			return $send_back;
 		}
 
-		foreach ( $users as $user_id ) {
-			$user = new \WP_User( $user_id );
-			if ( 'add_activitypub_cap' === $action ) {
+		// Handle adding ActivityPub capability.
+		if ( 'add_activitypub_cap' === $action ) {
+			foreach ( $users as $user_id ) {
+				$user = new \WP_User( $user_id );
 				$user->add_cap( 'activitypub' );
-			} elseif ( 'remove_activitypub_cap' === $action ) {
+			}
+			return $send_back;
+		}
+
+		// Handle removing ActivityPub capability - remove capabilities immediately, then redirect to fediverse deletion confirmation.
+		if ( 'remove_activitypub_cap' === $action ) {
+			$removed_count = 0;
+
+			// Remove capabilities immediately.
+			foreach ( $users as $user_id ) {
+				$user = new \WP_User( $user_id );
+
+				// Check if user has ActivityPub capability.
+				if ( ! $user->has_cap( 'activitypub' ) ) {
+					continue;
+				}
+
+				// Remove the capability.
 				$user->remove_cap( 'activitypub' );
+
+				// Force cache refresh for user capabilities.
+				\wp_cache_delete( $user_id, 'users' );
+				\wp_cache_delete( $user_id, 'user_meta' );
+
+				++$removed_count;
+			}
+
+			// Build the query args with proper array handling for fediverse deletion confirmation.
+			$query_args = array(
+				'activitypub_confirm_removal' => '1',
+				'send_back'                   => rawurlencode( $send_back ),
+			);
+
+			// Add user IDs as separate parameters.
+			foreach ( $users as $index => $user_id ) {
+				$query_args[ 'users[' . $index . ']' ] = absint( $user_id );
 			}
+
+			$confirmation_url = \add_query_arg( $query_args, \admin_url( 'users.php' ) );
+
+			// Force redirect instead of just returning URL.
+			\wp_safe_redirect( $confirmation_url );
+			exit;
+		}
+
+		// This should be handled by handle_bulk_capability_removal_confirmation
+		// but we include it here as a fallback.
```

Comment:
```suggestion
		// This should be handled by handle_bulk_capability_removal_confirmation but we include it here as a fallback.
```
Minor: Let's keep it on a single line or use [multiline comment docs](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#5-2-multi-line-comments).